### ISSUE TITLE
[CLI]: wandb.finish() stuck when uploaded all data

### ISSUE THREAD

**Issue by imoneoi** [wandb/wandb#4929]
### Describe the bug

When running a training loop multiple times and calling `wandb.finish()` after each run, although it shows that all data is uploaded, the program is still stuck for a very long time.

```python
def run_multiple_times():
    while True:
        wandb.init(reinit=True, ...)
        # training code...
        wandb.finish()


```

```shell
wandb: Waiting for W&B process to finish... (success).
wandb: | 20.180 MB of 20.180 MB uploaded (0.000 MB deduped)
```

### Environment

WandB version: `0.13.9`

OS: `5.4.0-135-generic #152-Ubuntu`

Python version: `3.10.9`

Versions of relevant libraries:

**Comment by MBakirWB** [comment]
Hi @imoneoi, happy to help. Could you please provide the following:
- `debug.log` and `debug-internal.log` for the handing runs
- How many runs are you executing? 
- Are you uploading any files/creating artifacts with these runs? If yes what is their approximate size?

**Comment by tfriedel** [comment]
I'm also seeing this issue. It doesn't happen everytime, but quite frequently. I have to kill wandb-service to make it disappear. 
I was running like 20 runs sequentiall and it would happen in maybe 20% of them?
Files being uploaded are rather small, like 250kb i.e. only code diffs, no models. Network connection is also pretty good (datacenter with gigabit+). 
I can send you log files privately if you tell me where.

**Comment by imoneoi** [comment]
I've found a temporary solution by using `wandb.init(reinit=True, ...)` and not calling `wandb.finish()` after a run.

My observation is similar to @tfriedel , about 20% of the runs will stuck.

- I started 10 runs in parallel (by launching different processes from terminal)
- Each run will upload 20 files with ~1MB each

**Comment by joinvalgo** [comment]
If the log output is bothering you, press `Ctrl+pq` in order to stop that message. It will disconnect the terminal from the container output.

**Comment by pilot7747** [comment]
I am also experiencing this issue, which makes it impossible to use sweeps because the runs get stuck on a wandb sync. In my case, this also results in extremely long syncs during the run (not just at the end), and sometimes the workspace does not update for 10-20 minutes. I am using `wandb` version 0.14.0, and I am not having any issues with my internet connection.

**Comment by MBakirWB** [comment]
@pilot7747 , could you please expand on what you are experiencing and your experiment setup. Is it that you ran a sweep and the sweep run was in a hung state and didn't finish successfully, or is it that you killed a run from the terminal and wandb  never closed out the run. If you have any run debug log files, that would be very helpful for review. In your wandb working directory `wandb>run-folder>logs`.

**Comment by pilot7747** [comment]
@MBakirWB Sweep run didn't finished and stuck at sync, so it couldn't start a new sweep run
I'll send the logs when I make a new try. At the first glance, running in a docker container solves the problem, so maybe it's some packages/OS issue.

**Comment by MBakirWB** [comment]
Hi @pilot7747, following up to see if you've been able to replicate the issue again and if you've had any logs for us to review?

**Comment by JacobiSong** [comment]
I have met a similar issue.
WandB version: 0.14.2.
Python: 3.10.10
OS: CentOS7
## log

<img width="583" alt="截屏2023-04-14 11 40 16" src="https://user-images.githubusercontent.com/61410169/231935723-6667c400-9723-40e7-b23e-54158c90369b.png">

## debug.log

<img width="978" alt="截屏2023-04-14 11 44 57" src="https://user-images.githubusercontent.com/61410169/231936420-61f4c3c2-74d1-46c8-9e47-5c709bd2e3f1.png">


## debug-internal.log

<img width="919" alt="截屏2023-04-14 11 37 01" src="https://user-images.githubusercontent.com/61410169/231935305-b212f8ce-7941-4300-a731-8877364ad08e.png">

**Comment by MBakirWB** [comment]
Hi @JacobiSong , could you provide details on how you got to this state?

**Comment by MBakirWB** [comment]
Hi @JacobiSong , since we have not heard back from you we are going to close this request. If you would like to re-open the conversation, please let us know!

**Comment by basicskywards** [comment]
Encountered the same issue and try some steps as below:

1. Typing ```top -u myusername``` in the command line will show the PID ```1754207``` for  ```wandb-service``` (you may have multiple wandb-service, so assume this PID causes the issue) like figure below
![image](https://user-images.githubusercontent.com/39130870/234491584-9b61f5fa-98dc-4ad3-98b5-5eb1fdadb223.png)

2. Typing ```kill -9 1754207``` to stop this wandb-service process

Problem luckily solved.

Reference here: https://stackoverflow.com/questions/54752710/nfs-file-cant-be-removed-resource-busy-but-pid-unknown
![image](https://user-images.githubusercontent.com/39130870/234493250-1b60ff34-1942-4fdd-848a-7a1054fd7c08.png)

**Comment by EvanBrownVTM** [comment]
Same issue

**Comment by Gutianpei** [comment]
I got the same issue when logging images with torch ddp training. Have to manually find the pid of wandb process and kill it to make the CLI clear again.

**Comment by nRknpy** [comment]
I had the same issue when logging matplotlib.Figure with wandb.log(). Although it is not an essential solution, the following method automatically kills the wandb process.

```
# This function forcibly kills the remaining wandb process.
def force_finish_wandb():
    with open(os.path.join(os.path.dirname(__file__), '../wandb/latest-run/logs/debug-internal.log'), 'r') as f:
        last_line = f.readlines()[-1]
    match = re.search(r'(HandlerThread:|SenderThread:)\s*(\d+)', last_line)
    if match:
        pid = int(match.group(2))
        print(f'wandb pid: {pid}')
    else:
        print('Cannot find wandb process-id.')
        return
    
    try:
        os.kill(pid, signal.SIGKILL)
        print(f"Process with PID {pid} killed successfully.")
    except OSError:
        print(f"Failed to kill process with PID {pid}.")

# Start wandb.finish() and execute force_finish_wandb() after 60 seconds.
def try_finish_wandb():
    threading.Timer(60, force_finish_wandb).start()
    wandb.finish()

# trainning scripts

# use try_finish_wandb instead of wandb.finish
try_finish_wandb()
```

**Comment by EvanBrownVTM** [comment]
I believe the issue is slow upload speed, if you give it a couple hours the
process should finish on it's own

**Comment by bhatiaabhinav** [comment]
Same problem.

**Comment by TalhaUusuf** [comment]
same problem with accelerate 4 gpus

**Comment by yt4766269** [comment]
same problem here

**Comment by tdye24** [comment]
> ```python
> reinit
> ```

Same issue as you!!!

**Comment by tdye24** [comment]
> I am also experiencing this issue, which makes it impossible to use sweeps because the runs get stuck on a wandb sync. In my case, this also results in extremely long syncs during the run (not just at the end), and sometimes the workspace does not update for 10-20 minutes. I am using `wandb` version 0.14.0, and I am not having any issues with my internet connection.

Same issue as yours!!!

**Comment by pilot7747** [comment]
I assume it might be a problem on W&B backend side, I'm not experiencing it anymore and there is nothing in the logs, so it seems that the process is just stuck waiting for something from the server.

**Comment by Lily-Le** [comment]
Same problem here. But I'm running multiple exps with wandb. Simpling k
illing wandb-service may influence other runnings. How to solve it?

**Comment by alexanderswerdlow** [comment]
Same problem for me. Very frustrating as I've killed a few training runs by accident at this point trying to kill the right wandb process.

**Comment by spfrommer** [comment]
Same issue.

**Comment by ericxsun** [comment]
I got the same issue and I found this error in `debug-internal.log`
```
WARNING FileStreamThread:193 [file_stream.py:request_with_retry():668] requests_with_retry encountered retryable exception: 500 Server Error: Internal Server Error for url: http://xxxx.com:8080/files/llm/project_name/exp_name/file_stream. func: functools.partial(<bound method Session.post of <requests.sessions.Session object at 0x7fb5be22fdc0>>, timeout=180),...
```

So how to solve it? Thanks a lot.

**Comment by swyoon** [comment]
Same issue here.

**Comment by realjoenguyen** [comment]
+1 got the same issue

**Comment by zhangjingyun7** [comment]
I have the same issue, and it happens more frequently now. Is there a better solution?  It's too cumbersome and a waste of time to killing the wandb-service manually  each time.

**Comment by Tendo33** [comment]
> I have the same issue, and it happens more frequently now. Is there a better solution? It's too cumbersome and a waste of time to killing the wandb-service manually each time.我也有同样的问题，而且现在出现的频率更高了。有更好的解决方案吗？每次都手动杀掉wandb-service太麻烦而且浪费时间。

exactly！！！

**Comment by zhangjingyun7** [comment]
I have tried to set the wandb on "offline" mode, and the issue disappeared. But there  arisen a new problem that the runs always repeat which means that when I have 100 runs, only about 50 are effective.

**Comment by huweiATgithub** [comment]
same problem. During the training, the metrics are uploaded to wandb's server without issues. When `wandb.finish()` is called, program stuck.

**Comment by melang982** [comment]
Same here

**Comment by JRamos1993** [comment]
Same issue.

**Comment by guangnianyuji** [comment]
Same issue.

**Comment by ruanvdmerwe** [comment]
Same issue

**Comment by realjoenguyen** [comment]
while waiting for the official fix, here is my script to quickly kill `wandb-service` 
`#!/bin/bash

USERNAME=nguyejoe
PATTERN=wandb-service
pgrep -u $USERNAME -f "^$PATTERN" | while read PID; do
    echo "Killing process ID $PID"
    kill $PID
done
`

**Comment by oliver-batchelor** [comment]
This

**Comment by Emiyalzn** [comment]
same issue

**Comment by Yang-Xijie** [comment]
same issue...

**Comment by RichealYoung** [comment]
same issue

**Comment by zzbuzzard** [comment]
same issue

**Comment by muzhi1991** [comment]
how to fix it? when use sweep, this will block next run.

**Comment by bluelancer** [comment]
same issue

**Comment by aradhyamathur** [comment]
I think the same issue is also pointed out here [](https://github.com/wandb/wandb/issues/5244)

**Comment by aradhyamathur** [comment]
One slightly better way i identified is to kill the wandb process that is not taking much resources, and is generally at lower priority with fewer threads, seemed to have worked out for me almost always when killing.

**Comment by kalomaze** [comment]
please fix this.

**Comment by AntiLibrary5** [comment]
same issue. And i was remotely logged into server!

**Comment by alexanderswerdlow** [comment]
I think I might have found a better solution that _should_ avoid killing other wandb/training processes (although obviously still use with care!).

Instead of just killing all wandb-processes, this first finds processes with the stdout file descriptor for the current shell (`$$` returns the shell PID) and _then_ filters for a wandb process owned by the current user.

Hope it helps!

```
alias kw='lsof -t "/proc/$$/fd/1" | xargs -I {} ps -u -f {} | grep "$USER" | grep "wandb-service" | grep -v grep | awk '\''{print $2}'\'' | xargs kill -9'
```

**Comment by davidsvaughn** [comment]
same issue here... first noticed this 3-4 years ago... now I'm seeing it again...
pssss.... wandb people.... maybe look into this?  ¯\_(シ)_/¯

**Comment by mvmacfarlane** [comment]
same here

**Comment by fireblue95** [comment]
same issue

**Comment by realjoenguyen** [comment]
Wandb new core doesn't seem to have this issue. See how to update in here 
[https://github.com/wandb/wandb/blob/main/core/README.md](https://github.com/wandb/wandb/blob/main/core/README.md)

**Comment by fireblue95** [comment]
@realjoenguyen 
"It works! Thank you for your solution."

**Comment by mvmacfarlane** [comment]
https://github.com/wandb/wandb/issues/5244#issuecomment-1666349612

**Comment by WhenMelancholy** [comment]
same issue here and this problem is really frequent.

**Comment by luisbergua** [comment]
Hello all,

Luis from the Engineering Support Team here. First of all, please accept our sincere apologies for the radio silence. Unfortunately, we missed the notifications for this thread, which caused the delay in our responses.

A big thank you to @realjoenguyen for already suggesting the solution! We've recently released our new backend, wandb-core (details [here](https://github.com/wandb/wandb/blob/main/core/README.md)), which is now the default in version 0.18.0. This update should resolve the issue you've been facing, alongside other major improvements like significantly faster processing, so I encourage you all to install the latest version.

I've enabled email alerts for this thread moving forward, so please reach out to us either here or via support@wandb.com if you still experience the issue after upgrading, and we'll investigate. We truly appreciate all the feedback and look forward to continuing to improve the tool with your help, so please keep it going!

**Comment by jonathanking-absci** [comment]
I'm still experiencing this in `wandb-0.18.7`. Am I think only one?

**Comment by luisbergua** [comment]
Hey @jonathanking-absci, are you able to reproduce this consistently? If so, would you have any problems with sharing an example? I tried different ways, but was unable to reproduce the issue. We added some improvements, but it might be that there are still some edge cases affecting here

**Comment by aradhyamathur** [comment]
It still occurs at times quite inconsistently, usually when we kill a process suddenly using interrupt (generally noticed when I kill some process using keyboard interrupt Ctrl+c ) , though the frequency of this occurring has reduced a lot in the past updates it still tends to occur at random sometimes. Havent been able to reproduce it either.

**Comment by jonathanking-absci** [comment]
@luisbergua Apologies, I later learned that my employer has a wandb slack channel where I brought this up. I can update my response here when we gain a better understanding of the issue. Yes, when running many of the same kinds of jobs simultaneously over a partitioned task, this happens consistently. However, trying to reproduce this using any individually partitioned job/task is not reliable.

**Comment by timoffex** [comment]
For those still affected, consider upgrading to version 0.19.1 of `wandb` which shows more information during `run.finish()` (ensure you're not using `wandb.require("legacy-service")`). This may surface network-related issues which are preventing the run from finishing, if that happens to be the problem.

**Comment by siwuxei** [comment]
me too
